### ISSUE TITLE
Improve fuzzing tests

### ISSUE THREAD

**Issue by kaleidawave** [kaleidawave/ezno#109]
The fuzzing of the parser is really good 🎉 some of the errors found in the parser so far:
- Array elements != function arguments. Array elements are allowed holes (e.g. `[2,,3]` is valid), but function arguments are not.
- When rendering `x++ + 5` it is really important to have the whitespace between the `++ +` as `x+++5` is invalid and has ambiguous parsing.
- If there is `(function () { return 2 })` in statement position, the parenthesis need to be kept as otherwise it is parsed as a statement and it is invalid as it needs a name (similarly for object literals and class expression).
- Lots of others...

---

**However** the fuzzing tests are still not passing. They are only really picking up extreme edge cases now , so it isn't urgent. I am hoping the fixes are near the end. But I was wondering if there could be some improvements:

- When a `parse -> render -> parse` (second parse) fail happens it shows `input` as what is mismatched. It is possible to print the first output as where the mismatch was. [See action run here](https://github.com/kaleidawave/ezno/actions/runs/7574254711/job/20628261352?pr=107#step:8:1237), I got confused at first as I thought the test was expected to match the input exactly (which is very difficult as it is just an AST parser)
- If there is a fail on either of the tests is it possible to keep going and maybe pick up say three or more errors. At the moment it short circuits on the first fail. Wondering if it is possible to gather more errors to fix as at the moment attempting to parse them is a very slow trial & commit. I couldn't find the input arguments in the docs: [Guide - Rust Fuzz Book (rust-fuzz.github.io)](https://rust-fuzz.github.io/book/cargo-fuzz/guide.html#libfuzzer-configuration-options), don't know if it is customisable. Even if it is just temporary for getting the first green light?
- I was wondering whether the fetching of boa's fuzzing cases in `build.rs` should get from a specific fixed ref, rather than the main branch?

Pinging @addisoncrump & @jasikpark *the fuzzing experts*

**Comment by addisoncrump** [comment]
_experts_ is strong :wink: 

1. `parse -> render -> parse`: Is the concern about the clarity of output? This can be adjusted quite easily, but what that is saying is that the input is parse-able, but the rendered output is not.
2. Picking up more errors is possible, but potentially not desirable. You may have multiple outputs which represent different bugs (or potentially a linear combination of bugs!). The relevant CLI is: `cargo fuzz run ... -- -fork=1 -ignore_crashes=1`
3. They should probably be fixed, yes :joy: Random breakages are never fun for CI.

**Comment by kaleidawave** [comment]
Awesome, I will look into that and try those out in a new PR. Do the fuzzing tests use libFuzzer and is that where the options are documented? https://llvm.org/docs/LibFuzzer.html#options

**Comment by kaleidawave** [comment]
Ah exactly what I wanted! More cases printed 🎉: https://github.com/kaleidawave/ezno/actions/runs/7713102547/job/21022255510 good for fixing the current issues, then can go back to short-circuited version.

**Comment by addisoncrump** [comment]
If you want to be a bit experimental: try the following replacement in fuzz/Cargo.toml:
```diff
diff --git a/parser/fuzz/Cargo.toml b/parser/fuzz/Cargo.toml
index 2abb8b9..7caea01 100644
--- a/parser/fuzz/Cargo.toml
+++ b/parser/fuzz/Cargo.toml
@@ -18,7 +18,7 @@ boa_ast = { git = "https://github.com/boa-dev/boa.git", features = [
 boa_interner = { git = "https://github.com/boa-dev/boa.git", features = [
   "arbitrary",
 ] }
-libfuzzer-sys = "0.4"
+libfuzzer-sys = { git = "https://github.com/AFLplusplus/LibAFL.git", branch = "libfuzzer-best", package = "libafl_libfuzzer" }
 pretty_assertions = "1.3.0"
 proc-macro2 = "1.0.66"
```

This uses the [libafl_libfuzzer](https://github.com/AFLplusplus/LibAFL/blob/main/libafl_libfuzzer/Cargo.toml) which we have since stabilised. It'll spit out some maybe undesirable backtraces, but it finds things _way_ faster than libfuzzer.

**Comment by kaleidawave** [comment]
Is it a bug that `==2546933== ERROR: libFuzzer: deadly signal` are printed but they do not print at the end?

![image](https://github.com/kaleidawave/ezno/assets/26967284/b15dcddc-def2-4f5f-bc36-2df02d8fec9f)

Also what does each part of this tracing line mean?
```
#2202576: cov: 7616 ft: 29739 corp: 7037 exec/s 4639 oom/timeout/crash: 0/0/2 time: 496s job: 31 dft_time: 0
```
Does one of these numbers display how many fuzzing runs it has completed?

**Comment by kaleidawave** [comment]
Nice! so `job: 31` means it is currently running 31 in parallel?

---

> The ==2546933== ERROR: libFuzzer: deadly signal indicate that the fuzzer found an input which triggered a "deadly" signal (e.g. a SIGABRT or SIGSEGV). This just means that the fuzzer is doing it's job -- not a bug.

That is what I thought. Do you know why they are not logged at the end of the test then? (it is under the `-fork=1 -ignore_crashes=1` mode for 10 minutes btw)
https://github.com/kaleidawave/ezno/actions/runs/7989637543/job/21816634859

**Comment by addisoncrump** [comment]
> Nice! so `job: 31` means it is currently running 31 in parallel?

No, it means that it's launched 31 individual instances. You only ever have one running at a time, but libfuzzer may stop "jobs" and start new ones if e.g. there is persistent memory detected, if the job crashes, or for other mysterious internal reasons.

**Comment by addisoncrump** [comment]
It seems that `cargo-fuzz` only emits crash data [if the command is not deemed successful](https://github.com/rust-fuzz/cargo-fuzz/blob/main/src/project.rs#L386-L404); with `max_total_time`, the exit code is zero, so it's marked successful. Let me see if I have a workaround for this.

**Comment by addisoncrump** [comment]
Sadly no easy way around it, and it seems that the exit code is determined by the exit code of the last exited job -- which may or may not be 0.

However, you can add the following POSIX-compatible snippet to the end of your test to just dump the output:
```bash
if test -d fuzz/artifacts; then find fuzz/artifacts -type f -print -exec xxd {} \; -exec cargo fuzz fmt -s none module_roundtrip_structured {} \;; false; fi
```

**Comment by kaleidawave** [comment]
Ah I think I understand 👍 will try that in the future
